### ISSUE TITLE
Integration into Open Toonz

### ISSUE THREAD

**Issue by blurymind** [dalboris/vpaint#60]
Open toonz is a professional open source animation package which has been recently released. I would be interested to hear the opinion of Boris and any other developers involved in vpaint. 
It is almost on the same level as toonboom harmony, but it is open source and free for everyone to use/fork.
https://github.com/opentoonz/opentoonz

However, here is the opinion of a blender developer on the state of it's code at the moment:

> On the other hand they (opentoonz developers)...
> 
> - Have been slow at accepting pull requests (check their git page).
> - Have not made any statements for their plans for the open-source version (except that they will continue to work on a closed fork).
> - Have not been significant contributions from any of the original developers (from what I can see).
> - Have not made efforts to make the build process easy (readme was in Japanese for example).
> 
> And we know it was only opened because the company that bought them demanded it.
> 
> We will of course wait-and-see, so far this has all the hallmarks of a code-dump that won't receive much attention.

Open toonz can and will revolutionize the 2d animation industry - the way blender is doing with the 3d industry - and open doors to alot of indie studios - but that is not going to happen if it's left in it's current state and the open source community doesn't embrace it fully.
http://blenderartists.org/forum/showthread.php?394970-ToonzPremium-(2D-Animation-tool)-now-Open-Sourced/page5

Open Toonz is in a dire need of open source developers at the moment. 
- Do you think it would be possible to integrate vpaint into open toonz 
or
- Fork Opentoonz to use it as vpaint's editor, instead of reinventing the wheel and making an editor for vpaint from scratch - recreating all the features that open toonz already has. The licensing of the two applications seems to allow that. You can fork it and do whatever you like with the code.

The design of opentoonz seems to be very compatible with vpaint. Perhaps vpaint could be a new level type? 

Just throwing ideas out there :)
Mostly have opentoonz support saving to your file format.

**Comment by blurymind** [comment]
Btw opentoonz already has an inbetweening vector technology that is similar to vpaint's.
It takes into consideration the order of drawing and direction of lines.
Lines can be sculpted after drawn, they can be invinsible or visible and contain a value of varied thickness in each node.
Lines however can not be flattened. You can only merge the end nodes of lines.

If anything else, it is an interesting piece of software to look into.

**Comment by blurymind** [comment]
You can literally fork open toonz, put vpaint as a layer (level) type, expand the vec file format to include raster graphics - make it the default file type and rename the entire application to Vtoonz or Vpaint pro or something :D 

My head will explode just imagining what the two technologies can do together.

**Comment by dalboris** [comment]
Hi Todor! Yes, it is absolutely exciting to see Toonz released Open Source! I'm glad you brought the topic here. I was aware of the open sourcing of Toonz a few weeks ago, but didn't have the time to try it yet.

I didn't yet took the time to carefully consider all aspects of the question, but as a first answer, I think for now I'm not going to try to integrate VPaint into OpenToonz, or the other way around. Doing this would surely be a lot of work and I just don't have the time. I'd rather keep VPaint separate for now, move forward the technology, and only when VPaint is mature enough, maybe we can reconsider and see what are our options.

For now, there are still many important open problems to solve for the Vector Graphics Animation technology to be production-ready, and I'd rather tackle these problems within a simpler dedicated app, VPaint, than within a larger app which is just going to bring many more problems and slow down development. In other words, I think that for now, resources are better spend on research on the technology rather than engineering work. It's way too early to commit to heavy engineering. A bit like this major Kickstarter failure, who commited to stock before ensuring that the prototype was actually working (http://www.digitaltrends.com/cool-tech/zano-drone-maker-says-sorry-for-failed-kickstarter-project/). First, we need to get it right ;-)

**Comment by scribblemaniac** [comment]
We'll since you invited other dev options, I'll quote my rant from elsewhere (you've already touched on many of the main points):

> I don't really want to go anywhere near that project at the moment. Between the people making tons of changes to get it to compile on linux, to those coming up with issue after issue about how seemingly nothing works on the currently supported operating systems, to ui that is missing major components and looks like it's running on System 7, to the pull requests that are sitting around collecting dust, it's an absolute mess right now in my opinion and I want nothing to do with it until they get their stuff together.

Now for some specific feedback for your suggestion: no. I understand where you're coming from, but integrating into OpenToonz is *way* more difficult than you make it sound. It isn't feasible with our current base of developers, and the number of developers willing to help out with OpenToonz is likely falling. I've done a bit of looking through the OpenToonz source, and being completely honest, the code for VPaint is far better. It's better documented, it's better structured, and don't get me started on libtiff or the compilation process.

VPaint's UI is nothing spectacular, in fact it's probably below average because of the mixture of native/non-native theming. I'm guessing you're on Windows, because on Mac/Linux OpenToonz looks horrendous. I'm pretty sure I'd be trading an animation studio for an insane asylum if I had to work on that all day.

Having said all this, it's possible that the OpenVac component of VPaint could be integrated into OpenToonz. Their code isn't structured very well for this at the moment, but maybe some ambitious developers there could do it. Even if they did, I suspect there would be many features from VPaint that cannot easily be supported by OpenToonz. I believe that dalboris' intention when designing OpenVac was to have it integrated into other projects like this.

In summary, OpenToonz is pretty terrible right now, and it's platform may not be able to support some of VPaint's features, so I am against the proposal.

**Comment by dalboris** [comment]
Thanks for your opinion scribblemaniac, it's great that you actually looked at it more than I did! Thus definitely, with this additional info, I'm not going to consider integrating into OpenToonz any time soon.

My intention for OpenVac is indeed to make it easier to integrate in project like this, but I'd rather not do the integration myself. Instead, I want to focus on research on the technology, which is where my skills are better leveraged, and VPaint is the platform to do this research. Having written all (well, most ;-) ) of the code myself allows to experiment much more than I ever could if I was working within the bounds of another piece of software. Though, note that it's a delicate balance: I still want VPaint to be well engineered and have as many professional features as possible, since being able to create professional content is the only way to evaluate whether or not the technology actually works, and what are its limitations, and try to overcome them. Which is why I'm actually doing a heavy refactoring of VPaint right now to make this possible.

I guess we'll have a better idea of where things are going in two or three years, not before unfortunately.

**Comment by blurymind** [comment]
The linux version is not fully working at the moment.
I think you should not judge the software by it's default setup. The japanese developers made some illogical changes to the default interface (to suit japanese studios traditional pipeline, but break everyone else's) that leave a bad first impression on many of the users.

any of the shortcomings you listed can be fixed by customizing the interface. Until you try the software, you wont appreciate the potential that it has. 

Anyways, it was worth a shot to try to get you to look into the source code at least. See if there's anything that might be useful for vpaint in there.

**Comment by dalboris** [comment]
Yes, definitely! As inspiration for VPaint features, I think OpenToonz is an invaluable resource!

**Comment by dalboris** [comment]
By the way, I appreciate that they released it in a BSD-style license, so indeed, if useful, VPaint can actually use its code.

**Comment by scribblemaniac** [comment]
> Until you try the software, you wont appreciate the potential that it has

I do believe it has potential, but it has a long ways to go. UX is huge, and it doesn't seem that the original team sees the need to change it.

> Yes, definitely! As inspiration for VPaint features, I think OpenToonz is an invaluable resource!

Yes for sure.

> By the way, I appreciate that they released it in a BSD-style license, so indeed, if useful, VPaint can actually use its code.

Good to hear, I was just about to ask that.

**Comment by blurymind** [comment]
The theming should not be too difficult to fix. It is using the same GUI library as vpaint. :)

Please have a look at toonz harlequin- the software that they forked it from:
https://youtu.be/xOzRNhTpGDs
![o3](https://cloud.githubusercontent.com/assets/6495061/14229267/7557c2cc-f927-11e5-9280-633fac24ee2a.png)

This is how it looked like before Dwango started messing with it's source code

**Comment by scribblemaniac** [comment]
If they were actually using qt properly this wouldn't be an issue. Well that does look better, but there are still some ways to go. Maybe someone will make a fork that catches on, maybe the original developers will start to take more notice, maybe it will just fall apart, only time will tell.

**Comment by justinclift** [comment]
The Blender person who said OpenToonz has the classic signs of a code-dump that won't get much attention, so far is correct.

The first few days after it's release saw a *significant* number of people trying it out (from the English speaking forums).  Kind of guessing that may have caught them off guard.  Hopefully they've noticed that, and will meaningfully encourage the development side of things & not blow it. :wink:

**Comment by dalboris** [comment]
I'm closing this for now, as I have no plans to integrate into OpenToonz considering the various arguments which have been mentioned. We can reopen in a few years, if relevant. Thanks all for your input!
